include/vara/Feature/FeatureModel.h

```diff
@@ -1,96 +1,206 @@
+  LLVM_DUMP_METHOD
+  void dump() const;
 
-    Feature *operator*() const { return It->second.get(); }
+protected:
+  string Name;
+  fs::path Path;
+  FeatureMapTy Features;
+  Feature *Root;
 
-    FeatureModelIter &operator++() {
-      It++;
-      return *this;
-    }
+private:
+  OrderedFeatureTy OrderedFeatures;
+};
 
-    bool operator==(const FeatureModelIter &Other) const {
-      return It == Other.It;
-    }
+//===----------------------------------------------------------------------===//
+//                     Builder for FeatureModel
+//===----------------------------------------------------------------------===//
 
-    bool operator!=(const FeatureModelIter &Other) const {
-      return not operator==(Other);
-    }
-  };
+/// \brief Builder for \a FeatureModel which can be used while parsing.
+class FeatureModelBuilder : private FeatureModel {
+public:
+  FeatureModelBuilder() : FeatureModel("", "", {}, nullptr){};
+
+  void init() {
+    Name = "";
+    Path = "";
+    Root = nullptr;
+    Features.clear();
+    Constraints.clear();
+    Parents.clear();
+    Children.clear();
+    Excludes.clear();
+  }
 
-  FeatureModelIter begin() { return FeatureModelIter(Features.begin()); }
+  /// Try to add \a BinaryFeature.
+  bool addFeature(const std::string &Key, bool Opt = false,
+                  std::optional<FeatureSourceRange> Loc = std::nullopt) {
+    return Features
+        .try_emplace(Key,
+                     std::make_unique<BinaryFeature>(Key, Opt, std::move(Loc)))
+        .second;
+  }
 
-  FeatureModelIter end() { return FeatureModelIter(Features.end()); }
+  /// Try to add \a NumericFeature.
+  bool addFeature(const std::string &Key,
+                  const NumericFeature::ValuesVariantType &Values,
+                  bool Opt = false,
+                  std::optional<FeatureSourceRange> Loc = std::nullopt) {
+    return Features
+        .try_emplace(Key, std::make_unique<NumericFeature>(Key, Values, Opt,
+                                                           std::move(Loc)))
+        .second;
+  }
 
-  LLVM_DUMP_METHOD
-  void dump() const {
-    for (const auto &F : this->Features) {
-      F.second->dump();
-      llvm::outs() << "\n";
-    }
-    for (auto C : this->Constraints) {
-      for (auto F : C) {
-        if (!F.second) {
-          llvm::outs() << "!";
-        }
-        llvm::outs() << F.first->getName();
-        if (F != C.back()) {
-          llvm::outs() << " | ";
-        }
-      }
-      llvm::outs() << "\n";
-    }
+  bool addFeature(Feature &F);
+
+  FeatureModelBuilder *addParent(const std::string &Key,
+                                 const std::string &ParentKey) {
+    Children[ParentKey].insert(Key);
+    Parents[Key] = ParentKey;
+    return this;
+  }
+
+  FeatureModelBuilder *addExclude(const std::string &Key,
+                                  const std::string &ExcludeKey) {
+    Excludes[Key].insert(ExcludeKey);
+    return this;
+  }
```

Comment:
What happens in these functions if the `Key` is not a Feature name that was inserted before?

PS: `Key` is also a bad name for this param, we should rename this.

Reply:
Doing so is totally fine, as the consistency (existence) is asserted when building the final FM. As the parser can not assure that features are only referenced if they are already processed (like mutual exclusion), this can not be asserted in this method either.